File: native-migration/slices/01-order-tracking/ios-native/TrackingApi.swift
import Foundation
import Combine

/// Phase 2 Native Slice: Order Tracking API
/// Contract-compatible with BhojanOS Marketplace API.
protocol TrackingApiProtocol {
    func getTracking(orderId: String) -> AnyPublisher<OrderTrackingResponseDTO, Error>
    func getGuestTracking(orderId: String, phone: String) -> AnyPublisher<OrderTrackingResponseDTO, Error>
}

class TrackingApi: TrackingApiProtocol {
    private let baseURL = URL(string: "https://api.bhojanos.com/api/marketplace")!
    private let urlSession: URLSession
    
    init(urlSession: URLSession = .shared) {
        self.urlSession = urlSession
    }
    
    func getTracking(orderId: String) -> AnyPublisher<OrderTrackingResponseDTO, Error> {
        let url = baseURL.appendingPathComponent("orders/\(orderId)/tracking")
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        // TODO: Inject Authorization header from session manager
        
        return urlSession.dataTaskPublisher(for: request)
            .map(\.data)
            .decode(type: OrderTrackingResponseDTO.self, decoder: JSONDecoder())
            .eraseToAnyPublisher()
    }
    
    func getGuestTracking(orderId: String, phone: String) -> AnyPublisher<OrderTrackingResponseDTO, Error> {
        var components = URLComponents(url: baseURL.appendingPathComponent("orders/\(orderId)/guest-tracking"), resolvingAgainstBaseURL: false)!
        components.queryItems = [URLQueryItem(name: "phone", value: phone)]
        
        var request = URLRequest(url: components.url!)
        request.httpMethod = "GET"
        
        return urlSession.dataTaskPublisher(for: request)
            .map(\.data)
            .decode(type: OrderTrackingResponseDTO.self, decoder: JSONDecoder())
            .eraseToAnyPublisher()
    }
}

// MARK: - Data Transfer Objects (from src/types/marketplace.ts)

struct OrderTrackingResponseDTO: Codable {
    let orderId: String
    let orderNumber: String
    let status: String
    let paymentStatus: String?
    let expiresAt: String?
    let timeline: [TrackingTimelineEventDTO]
    let etaMinutes: TrackingEtaDTO?
    let restaurant: TrackingRestaurantDTO?
    let delivery: TrackingDeliveryDTO?
}

struct TrackingTimelineEventDTO: Codable {
    let status: String
    let at: String
    let message: String?
}

struct TrackingEtaDTO: Codable {
    let min: Int
    let max: Int
}

struct TrackingRestaurantDTO: Codable {
    let displayName: String
    let slug: String
    let restaurantId: String
}

struct TrackingDeliveryDTO: Codable {
    let partner: String?
    let trackingUrl: String?
    let riderName: String?
    let riderPhone: String?
}

File: native-migration/slices/01-order-tracking/android-native/TrackingApi.kt
package com.bhojanos.app.tracking.data

import retrofit2.http.GET
import retrofit2.http.Path
import retrofit2.http.Query
import retrofit2.Response

/**
 * Phase 2 Native Slice: Order Tracking API
 * Contract-compatible with BhojanOS Marketplace API.
 */
interface TrackingApi {

    /**
     * Authenticated Tracking Endpoint
     * Headers (like Authorization) should be injected via OkHttp Interceptor
     * in the main Retrofit instance.
     */
    @GET("api/marketplace/orders/{orderId}/tracking")
    suspend fun getTracking(
        @Path("orderId") orderId: String
    ): Response<OrderTrackingResponseDTO>

    /**
     * Guest Tracking Endpoint (Requires Phone Number validation)
     */
    @GET("api/marketplace/orders/{orderId}/guest-tracking")
    suspend fun getGuestTracking(
        @Path("orderId") orderId: String,
        @Query("phone") phone: String
    ): Response<OrderTrackingResponseDTO>
}

// Data Transfer Objects matching `src/types/marketplace.ts`

data class OrderTrackingResponseDTO(
    val orderId: String,
    val orderNumber: String,
    val status: String,
    val paymentStatus: String? = null,
    val expiresAt: String? = null,
    val timeline: List<TrackingTimelineEventDTO>,
    val etaMinutes: TrackingEtaDTO? = null,
    val restaurant: TrackingRestaurantDTO? = null,
    val delivery: TrackingDeliveryDTO? = null
    // Ignoring invoice/feedback/reorder for now as tracking is the core scope
)

data class TrackingTimelineEventDTO(
    val status: String,
    val at: String,
    val message: String? = null
)

data class TrackingEtaDTO(
    val min: Int,
    val max: Int
)

data class TrackingRestaurantDTO(
    val displayName: String,
    val slug: String,
    val restaurantId: String
)

data class TrackingDeliveryDTO(
    val partner: String? = null,
    val trackingUrl: String? = null,
    val riderName: String? = null,
    val riderPhone: String? = null
)
